Adopt FinBuckle MultiTenant

fretje opened the pull request:
Fwiejuww.

It was quite some work, but got it mostly done.

Still working on it though...

I haven't tested a lot yet, and I know the initialization (when creating a new tenant) isn't working fully yet, but wanted to get it out here already, in case people would be interested in helping with testing a bit ;-)

I do know the blazor client works fine for a single tenant (needs a small update to the FSHApiClient).

fretje commented:
Initialization works, Seeding works, general users/roles/brand/product/dashboard management features works...

Still have to test hangfire... not immediately sure how to?

and also azureAd auth

No feedback yet? No reviews? I surely must have made some mistakes ;-)

274188A commented:
wow - quite a few changes - any chance of a summary?

fretje commented:
Well functionally it should behave exactly the same, but a lot of the complex multitenancy code is gone. It now relies on the FinBuckle MultiTenant library for that, which is quite impressive I have to say. ApplicationDbContext (or rather BaseDbContext) is now derived from MultiTenantIdentityDbContext (provided by the library). And you can configure a lot of stuff very easily at app startup... for instance which strategy you want to use to obtain the tenantId. I.e. BasePathStrategy if you want to work with subdomains, RouteStrategy, if your tenantId is in the route, headerstrategy, ... You can also configure which store you want to use. It's now default configured as a efcore store (a table in the database), but it can as well be in memory, or coming from configuration, ...
Anyways too much to tell... Everything is quite nicely documented:
https://www.finbuckle.com/MultiTenant/Docs/v6.5.1

fretje commented:
I have taken multitenancy out of domain and put it just like identity into infrastructure (mutlitenancy is even more "infrastructury" than identity if you ask me).

What used to be the `Tenant` entity is now called `FSHTenantInfo` which implements `ITenantInfo` (which is a FinBuckle type). You can inject an `ITenantInfo` or `FSHTenantInfo` (if you need more than the basic information) anywhere you would otherwise inject services, and it will contain the current tenant or null if none of the strategies configured encountered a TenantId in the request. It will probably throw right now when that is the case, but that's also configurable.

But in the end, as an application developer, you should never have to care about tenants... You're always inside the context of a tenant and everything about it is taken care of for you. You just code as if there are no tenants.

One thing you have to do though is telling FinBuckle that an entity is a multitenant entity. You can do that in the entityconfiguration with builder.IsMultiTenant(). See how it's done for brands and products now in the code of this pr...

We probably should find a way to automatically make all `BaseEntity`s multitenant... even less thing to care about for the developer ;-)

iammukeshm commented on src/Core/Application/Identity/Roles/RoleDto.cs:
```diff
@@ -5,6 +5,5 @@ public class RoleDto
     public string Id { get; set; } = default!;
     public string Name { get; set; } = default!;
     public string? Description { get; set; }
-    public string? Tenant { get; set; }
```
Why was this removed?

iammukeshm commented on src/Infrastructure/Multitenancy/TenantService.cs:
```diff
@@ -0,0 +1,111 @@
+
+    private async Task<FSHTenantInfo> GetTenantInfoAsync(string id) =>
+        await _tenantStore.TryGetAsync(id)
+            ?? throw new NotFoundException(string.Format(_localizer["entity.notfound"], typeof(FSHTenantInfo).Name, id));
```
tenant*.notfound

fretje replied:
that's what it says actually... it's using the generic entity.notfound which has 2 parameters, the name of the entitytype and the id... but yeah would probably be better with a specific one...

The localizer strings is also something that could use some "re-organising"  in general ;-)
and more consistency...

fretje replied:
I ran a search and there's no "tenant.notfound" used yet anywere.
Let's keep this for another pr... maybe should create an issue for it for another PR... I created an issue https://github.com/fullstackhero/dotnet-webapi-boilerplate/issues/372

fretje replied:
I might start to sound like a broken record, but I would personally go for the standard .NET way of working with resources. Which has a lot of tooling and support built around it...
Like you also do on the blazor project...

iammukeshm commented on src/Infrastructure/Multitenancy/TenantService.cs:
```diff
@@ -0,0 +1,111 @@
+
+        try
+        {
+            // TODO: we should probably run this in a hangfire job?
```
Yes, will save some thread load.

iammukeshm commented on src/Infrastructure/Notifications/NotificationHub.cs:
```diff
@@ -9,36 +10,35 @@ namespace FSH.WebApi.Infrastructure.Notifications;
     private readonly ILogger<NotificationHub> _logger;
 
-    public NotificationHub(ILogger<NotificationHub> logger)
+    public NotificationHub(ITenantInfo? currenTenant, ILogger<NotificationHub> logger)
```
currentTenant*

fretje replied:
wow sharp eye ;-) fixed

iammukeshm commented on src/Infrastructure/Persistence/Context/ApplicationDbContext.cs:
```diff
@@ -1,75 +1,20 @@
     public DbSet<Product> Products => Set<Product>();
     public DbSet<Brand> Brands => Set<Brand>();
-
-    public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken = default)
```
So this part is moved to BaseDbContext?

fretje replied:
damn I wrote a whole comment and made a mistake and it's gone again... but yes it's moved to basedbcontext, as it was already partly there. we should also take the 2 auditing methods out of basedbcontext into an entityauditingservice or something like that, but that's for another pr another day ;-)

fretje commented:
Everything seems to work... have also tested AzureAd and the GenerateBrands and DeleteBrands methods which use HangFire...

So Ready For Review ;-)

fretje commented:
Oh yes, concerning the connectionstrings... you can leave the connectionstring empty for tenants that use the same connectionstring as the root tenant... only if you want to have a separate database, you have to fill it in (even in the database... finbuckle takes care of that).

I think you can even leave the connectionstring for the root tenant empty, as the one configured in the settings will take precedence anyways...

fretje commented:
And another thing... I have regenerated the migration scripts for MSSQL... the other ones still need to be done... still haven't figured out how to do that...

I did make a small change in the folder structure... I renamed the `Root` folder to `Tenant`, so now it's `Application` and `Tenant`.

And I also added the tables in their respective schema's... Makes for a very nice DB layout:
![image](https://user-images.githubusercontent.com/775707/149603895-cc8d4b6d-178f-4bfc-be41-f96dba0b2d60.png)

We will just have to find a way to make the schema's uppercase in the case of Oracle... Although now I see this image... how is Oracle handling the Hangfire tables? Are they also uppercase on Oracle?

274188A commented:
I find the Oracle uppercase problem hard to fathom myself - I've NEVER heard of having to have uppercase and I have some experience with that DB

fretje commented:
Yeah have made it so it can all be change easily in one place:
```c#
internal static class SchemaNames
{
    // TODO: figure out how to capitalize these only for Oracle
    public static string Auditing = "Auditing"; // "AUDITING";
    public static string Catalog = "Catalog"; // "CATALOG";
    public static string Identity = "Identity"; // "IDENTITY";
    public static string MultiTenancy = "MultiTenancy"; // "MULTITENANCY";
}
```
People that want the uppercase can just change it here for now... I'd leave the default non uppercase...

fretje commented:
I just found a bug now when creating a new tenant with an emtpy Identifier it doesn't work... (identifier is defined as unique in the database).

I must have misinterpreted why that identifier is used exactly. I know it's the Id of the tenant that gets placed in the tenantId field of all entities. So I thought Identifier could just be used or not by the developer for something else... but apparently you can't leave it empty...

I've used the Identifier field now for storing the issuer when using azuread but I'll probably have to change that again... stay tuned ;-)

fretje commented:
Okay it's fixed... tested both creating a new tenant with and without connectionstring... and it seems to work fine... logged in and out of them and they have only their respective entities...
The error handling could be a bit better with regards to the connectionstring... but when there's no mistakes, the database creation and initialization works fine.

xlogex commented:
public DatabaseInitializer(TenantDbContext tenantDbContext, IMultiTenantStore<FSHTenantInfo> tenantStore, IOptions<DatabaseSettings> dbSettings, ...
    public async Task InitializeDatabasesAsync(CancellationToken cancellationToken)
    {
        // First initialize the tenant db context
        if ((await _tenantDbContext.Database.GetPendingMigrationsAsync(cancellationToken)).Any()) <-----------------add await ???

fretje commented:
@xlogex I see an await there?

Btw: you can actually add comments to individual code lines in the pr... That's what they're made for...

xlogex commented on src/Infrastructure/Persistence/Initialization/DatabaseInitializer.cs:
```diff
@@ -0,0 +1,81 @@
+    public async Task InitializeDatabasesAsync(CancellationToken cancellationToken)
+    {
+        // First initialize the tenant db context
+        if (_tenantDbContext.Database.GetPendingMigrations().Any())
```
if ((await _tenantDbContext.Database.GetPendingMigrationsAsync(cancellationToken)).Any())

iammukeshm replied:
This is not an Async method / Task, hence await is not needed.

DurgaPrasadReddyV replied:
I see many methods do not have async await decorators. Are we skipping use of await and async intentionally?

fretje replied:
> if ((await _tenantDbContext.Database.GetPendingMigrationsAsync(cancellationToken)).Any())

Oh now I see what you mean... unfortunately the pr is already merged... you can make one yourself for this ;-)

fretje replied:
> I see many methods do not have async await decorators. Are we skipping use of await and async intentionally?

When you're immediately returning a Task, it's not necessary to make the method async and then do await...
Or said in another way, when your method is async and it only uses await on the last method, it's better to take away the async and await...
(no need to set up an expensive state machine just to tear it down again)

iammukeshm commented:
> Yeah, I have made it so it can all be change easily in one place:
> 
> ```cs
> internal static class SchemaNames
> {
>     // TODO: figure out how to capitalize these only for Oracle
>     public static string Auditing = "Auditing"; // "AUDITING";
>     public static string Catalog = "Catalog"; // "CATALOG";
>     public static string Identity = "Identity"; // "IDENTITY";
>     public static string MultiTenancy = "MultiTenancy"; // "MULTITENANCY";
> }
> ```
> 
> People that want the uppercase can just change it here for now... I'd leave the default non uppercase...

About this, there should be a builder method or something which can handle schemas casing for Oracle. MySQL too had such an exception, since it doesn't support schemas via EFCore. Will check on that.

frankyjquintero commented:
> Encuentro el problema de las mayúsculas de Oracle difícil de entender por mí mismo: NUNCA he oído hablar de tener que tener mayúsculas y tengo algo de experiencia con ese DB

The problem is in the way entityframework performs a query to manage migrations. In the case of oracle, it has the novelty that since it is a schema, a user must first create each one of them. and in the case of capital letters it is due to a query that launches the oracle driver to search for tables and this is not done in a sensitive way but with a "select .... = 'IDENTITY' ". If you want to know more about this scenario you can use docker and these steps to replicate the process.

https://github.com/fullstackhero/dotnet-webapi-boilerplate/issues/225

frankyjquintero commented:
> And another thing... I have regenerated the migration scripts for MSSQL... the other ones still need to be done... still haven't figured out how to do that...
> 
> I did make a small change in the folder structure... I renamed the `Root` folder to `Tenant`, so now it's `Application` and `Tenant`.
> 
> And I also added the tables in their respective schema's... Makes for a very nice DB layout: ![image](https://user-images.githubusercontent.com/775707/149603895-cc8d4b6d-178f-4bfc-be41-f96dba0b2d60.png)
> 
> We will just have to find a way to make the schema's uppercase in the case of Oracle... Although now I see this image... how is Oracle handling the Hangfire tables? Are they also uppercase on Oracle?


This was tested on all supported database engines??

remember that not all engines use the schema concept in the same way and it can cause problems

fretje commented:
It's tested on all but Oracle... 

That's why I asked the question "How do the Hangfire tables end up in oracle?"... How do they (hangfire) handle it? Maybe we can employ something similar?

In any case, if it does give problems for people that use oracle, they only have to change the one file where the schemanames are defined...

frankyjquintero commented:
> It's tested on all but Oracle...
> 
> That's why I asked the question "How do the Hangfire tables end up in oracle?"... How do they (hangfire) handle it? Maybe we can employ something similar?
> 
> In any case, if it does give problems for people that use oracle, they only have to change the one file where the schemanames are defined...

HangFire does not have support for Oracle not directly, in this link https://www.hangfire.io/extensions.html you can see an option that uses FluentNHibernateStorage. But in any case of creating the user (because in oracle a schema is a user) it must be created first

frankyjquintero commented:
> I find the Oracle uppercase problem hard to fathom myself - I've NEVER heard of having to have uppercase and I have some experience with that DB

The error comes from: Oracle.ManagedDataAccess.Client

`
DECLARE
    USEREXIST INTEGER;
    USER_NOT_EXIST EXCEPTION;
    PRAGMA EXCEPTION_INIT( USER_NOT_EXIST, -01435 );
BEGIN
SELECT COUNT(*) INTO USEREXIST FROM ALL_USERS WHERE USERNAME='MultiTenancy';
IF (USEREXIST = 0) THEN
    RAISE USER_NOT_EXIST;
END IF;
END;
`

![image](https://user-images.githubusercontent.com/22311691/151068640-c455be81-4357-46ab-8141-87d8c532f8fb.png)

frankyjquintero commented:
Another problem with the separation of the schemas is the one exposed in this StackOverflow thread

https://stackoverflow.com/questions/63398380/database-creation-with-ef-core-code-first-migrations


Which forces to create each user for each schema first

iammukeshm commented:
is there a way to ignore schemas / schema creation if the db provider is Oracle? There is something similar with the EFCore extension for MySQL

```
return builder.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString), e =>
                     e.MigrationsAssembly("Migrators.MySQL")
                      .SchemaBehavior(MySqlSchemaBehavior.Ignore));
```

frankyjquintero commented:
from what I saw it does not have that option available, the idea of schema separation is not bad and in the end it is only a matter of permissions to create a user that can see the information of the schema users, or in which case the same is used schema from the configuration file that was created.
